docker-images/syntax-highlighter/crates/scip-syntax/src/locals.rs

```diff
@@ -1,504 +1,734 @@
+            self.add_defs_while(current_scope, &mut definitions_iter, |def_capture| {
+                def_capture.node.start_byte() < scope_end_byte
+            });
+            self.add_refs_while(current_scope, &mut references_iter, |ref_capture| {
+                ref_capture.node.start_byte() < scope_end_byte
+            });
 
-    let capacity = lvalues.len() + references.len();
+            if let Some(parent) = self.get_scope(current_scope).parent {
+                current_scope = parent
+            } else {
+                // We've made it to the top level scope
+                break;
+            }
+        }
 
-    // Add all the scopes to our tree
-    while let Some(m) = scopes.pop() {
-        root.insert_scope(m);
+        assert!(
+            definitions_iter.next().is_none(),
+            "Should've entered all definitions into the tree"
+        );
     }
 
-    while let Some(m) = lvalues.pop() {
-        root.insert_lvalue(m);
+    /// Walks up the scope tree and tries to find the definition for a given name
+    fn find_def(
+        &self,
+        scope: ScopeId<'a>,
+        name: Name,
+        start_byte: usize,
+    ) -> Option<&Definition<'a>> {
+        let mut current_scope = scope;
+        let mut fuel = MAX_SCOPE_DEPTH;
+        loop {
+            fuel -= 1;
+            if fuel <= 0 {
+                eprintln!("Detected a likely infinite loop in scip-syntax::locals::LocalResolver::find_def");
+                return None;
+            }
+            let scope = self.get_scope(current_scope);
+            if let Some(def) = scope.find_def(name, start_byte) {
+                return Some(def);
+            } else if let Some(parent_scope) = scope.parent {
+                current_scope = parent_scope
+            } else {
+                return None;
+            }
+        }
     }
 
-    root.clean_empty_scopes();
+    fn resolve_references(&mut self) {
+        let mut ref_occurrences = vec![];
+
+        for (scope_ref, scope) in self.arena.iter() {
+            for Reference {
+                name,
+                node,
+                resolves_to,
+            } in scope.references.iter()
+            {
+                let def_id = if let Some(resolved) = resolves_to {
+                    *resolved
+                } else if self.definition_start_bytes.contains(&node.start_byte()) {
+                    // See the comment on LocalResolver.definition_start_bytes
+                    continue;
+                } else if let Some(def) = self.find_def(scope_ref, *name, node.start_byte()) {
+                    def.id
+                } else {
+                    continue;
+                };
+
+                let symbol = format_symbol(def_id.as_local_symbol());
+                ref_occurrences.push(scip::types::Occurrence {
+                    range: node.to_scip_range(),
+                    symbol: symbol.clone(),
+                    ..Default::default()
+                });
```

Comment:
```suggestion
                ref_occurrences.push(scip::types::Occurrence {
                    range: node.to_scip_range(),
                    symbol: format_symbol(def_id.as_local_symbol()),
                    ..Default::default()
                });
```

nit: Prefer move over `clone()`.